fdbrpc/include/fdbrpc/FileTransfer.h

```diff
@@ -0,0 +1,134 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifdef FLOW_GRPC_ENABLED
+#ifndef FDBRPC_FILE_TRANSFER_H
+#define FDBRPC_FILE_TRANSFER_H
+
+#include <optional>
+#include <fstream>
+#include <grpcpp/grpcpp.h>
+
+#include <boost/asio.hpp>
+
+#include "fdbrpc/file_transfer/file_transfer.grpc.pb.h"
+
+using fdbrpc::DownloadChunk;
+using fdbrpc::DownloadRequest;
+using fdbrpc::FileTransferService;
+using grpc::Channel;
+using grpc::ClientContext;
+using grpc::Status;
+
+class FileTransferServiceImpl final : public FileTransferService::Service {
+public:
+	grpc::Status DownloadFile(grpc::ServerContext* context,
+	                          const DownloadRequest* request,
+	                          grpc::ServerWriter<DownloadChunk>* writer) override {
+		std::ifstream input_file(request->file_name(), std::ios::binary | std::ios::ate);
+		if (!input_file.is_open()) {
+			return grpc::Status(grpc::StatusCode::NOT_FOUND, "File not found");
+		}
+
+		const std::streamsize file_size = input_file.tellg();
+		const size_t buffer_size = request->chunk_size() > 0 ? request->chunk_size() : 1024 * 1024;
+		const size_t start_chunk_index = request->first_chunk_index();
+		int64_t offset = buffer_size * start_chunk_index;
+
+		if (offset > file_size) {
+			return grpc::Status(grpc::StatusCode::OUT_OF_RANGE, "Offset beyond file size");
+		}
+
+		input_file.seekg(offset);
+		if (!input_file) {
+			return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Failed to seek to the offset");
+		}
+
+		std::vector<char> buffer(buffer_size);
+
+		while (input_file.read(buffer.data(), buffer_size) || input_file.gcount() > 0) {
+			std::streamsize bytes_read = input_file.gcount();
+			DownloadChunk chunk;
+			chunk.set_offset(offset);
+			chunk.set_data(buffer.data(), bytes_read);
+			writer->Write(chunk);
+			offset += bytes_read;
+		}
+
+		if (input_file.bad()) {
+			return grpc::Status(grpc::StatusCode::DATA_LOSS, "Error while reading the file");
+		}
+
+		input_file.close();
+		return grpc::Status::OK;
+	}
+};
+
+class FileTransferClient {
+public:
+	FileTransferClient(std::shared_ptr<Channel> channel) : stub_(FileTransferService::NewStub(channel)) {}
+
+	std::optional<size_t> DownloadFile(const std::string& filename, const std::string& output_filename) {
+		ClientContext context;
+		DownloadRequest request;
+		request.set_file_name(filename);
+
+		std::unique_ptr<grpc::ClientReader<DownloadChunk>> reader(stub_->DownloadFile(&context, request));
+
+		std::ofstream output_file(output_filename, std::ios::binary | std::ios::trunc);
+		if (!output_file.is_open()) {
+			return std::nullopt; // Failed to open file
+		}
+
+		DownloadChunk chunk;
+		size_t bytes_read = 0;
+		while (reader->Read(&chunk)) {
```

Comment:
If exception, the partially written file is cleaned up?

Reply:
I was thinking, application may want to resume from last chunk it successfully fsynced.  But I think its a good point and need to documented properly. And I should add a parameter to set resume offset in local file.